File: Editorial/0138_Copy_List_with_Random_Pointer/CSharp/1_Recursive.cs
/*
// Definition for a Node.
public class Node {
    public int val;
    public Node next;
    public Node random;
    public Node(int _val, Node _next, Node _random) {
        val = _val;
        next = _next;
        random = _random;
    }
}
*/

public class Solution {
    // Dictionary which holds old nodes as keys and new nodes as its values.
    private Dictionary<Node, Node> visitedHash = new Dictionary<Node, Node>();

    public Node CopyRandomList(Node head) {
        if (head == null) {
            return null;
        }

        // If we have already processed the current node, then we simply return
        // the cloned version of it.
        if (this.visitedHash.ContainsKey(head)) {
            return this.visitedHash[head];
        }

        // Create a new node with the value same as old node. (i.e., copy the
        // node)
        Node node = new Node(head.val, null, null);
        // Save this value in the hash map. This is needed since there might be
        // loops during traversal due to randomness of random pointers and this
        // would help us avoid them.
        this.visitedHash[head] = node;
        // Recursively copy the remaining linked list starting once from the
        // next pointer and then from the random pointer. Thus we have two
        // independent recursive calls. Finally we update the next and random
        // pointers for the new node created.
        node.next = this.CopyRandomList(head.next);
        node.random = this.CopyRandomList(head.random);
        return node;
    }
}

File: Editorial/138_Copy_List_with_Random_Pointer/TypeScript/1_Recursive.ts
/**
 * Definition for Node.
 * class Node {
 *     val: number
 *     next: Node | null
 *     random: Node | null
 *     constructor(val?: number, next?: Node, random?: Node) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *         this.random = (random===undefined ? null : random)
 *     }
 * }
 */

function copyRandomList(head: Node | null): Node | null {
    // HashMap which holds old nodes as keys and new nodes as its values.
    let visitedHash: Map<Node, Node> = new Map();

    // Helper function to clone a node
    const cloneNode = function (node: Node | null): Node | null {
        if (node === null) {
            return null;
        }
        // If the node was already visited before.
        // Return the clone from the visited dictionary.
        if (visitedHash.has(node)) {
            return visitedHash.get(node)!;
        }
        // Create a new node
        // with the value same as old node.
        let newNode: Node = new Node(node.val, null, null);
        // Save this value in the hash map. This is needed since there might be
        // loops during traversal due to randomness of random pointers and this would help us avoid them.
        visitedHash.set(node, newNode);

        // Recursively copy the remaining linked list starting once from the next pointer and then from the random pointer.
        // Thus we have two independent recursive calls.
        // Finally we update the next and random pointers for the new node created.
        newNode.next = cloneNode(node.next);
        newNode.random = cloneNode(node.random);
        return newNode;
    };

    return cloneNode(head);
}
